thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/detection/DetectionJobScheduler.java

```diff
@@ -438,37 +437,33 @@ public void synchronousBackFill(long functionId, DateTime backfillStartTime, Dat
       return;
     }
 
+    int retryCounter = 0;
+    while (waitForJobDone(jobExecutionId).equals(JobStatus.FAILED) && retryCounter < MAX_BACKFILL_RETRY) {
+      jobExecutionId = runBackfill(functionId, backfillStartTime, backfillEndTime, force);
+      retryCounter++;
+    }
+  }
+
+  public JobStatus waitForJobDone(long jobExecutionId) {
     TaskManager taskDAO = DAO_REGISTRY.getTaskDAO();
     JobManager jobDAO = DAO_REGISTRY.getJobDAO();
-    JobDTO jobDTO = null;
+    JobDTO jobDTO;
     List<TaskDTO> scheduledTaskDTO = taskDAO.findByJobIdStatusNotIn(jobExecutionId, TaskConstants.TaskStatus.COMPLETED);
+    long functionId = jobDAO.findById(jobExecutionId).getAnomalyFunctionId();
 
-    int retryCounter = 0;
     while (scheduledTaskDTO.size() > 0) {
-      boolean hasFailedTask = false;
+      List<Long> failedTaskIds = new ArrayList<>();
       for (TaskDTO taskDTO : scheduledTaskDTO) {
         if (taskDTO.getStatus() == TaskConstants.TaskStatus.FAILED) {
-          if (retryCounter >= MAX_BACKFILL_RETRY) {
-            LOG.warn("The backfill of anomaly function {} (task id: {}) failed. Stop retry.", functionId, jobExecutionId);
-
-            // Set job to be failed
-            jobDTO = jobDAO.findById(jobExecutionId);
-            if (jobDTO.getStatus() != JobConstants.JobStatus.FAILED) {
-              jobDTO.setStatus(JobConstants.JobStatus.FAILED);
-              jobDAO.save(jobDTO);
-            }
-
-            return;
-          }
-          hasFailedTask = true;
-          LOG.warn("The backfill of anomaly function {} (task id: {}) failed. Retry count: {}", functionId,
-              jobExecutionId, retryCounter);
+          failedTaskIds.add(taskDTO.getId());
```

Comment:
If a task always failed, will this cause an infinite loop. Should think about ways to give a hard limit for failed jobs to restart